Implementing createImageBitmap

ramyananth opened the pull request:
<!-- Please describe your changes on the following line: -->
Implementing createImageBitmap method for canvas image source. 

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #20650 (GitHub issue number if applicable)

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

JayalakshmiV committed:
implemented CreateImageBitmap function for Canvas image source

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @ferjm (or someone else) soon.

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/window.rs, components/script/canvas_state.rs, components/script/dom/imagebitmap.rs
 * @kichjang: components/script/dom/window.rs, components/script/canvas_state.rs, components/script/dom/imagebitmap.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify script code, but no tests are modified. Please consider adding a test!

ramyananth commented:
Hi Josh, 
We have pushed out latest changes.
We get repetitive errors for the window.rs file and workerglobalscope.rs file. 
Error E0596 is thrown for every usage of the imageBitmap.set_origin_clean() and imageBitmap.set_bitmap_data() ie., [Link1](https://github.com/ramyananth/servo/blob/master/components/script/dom/window.rs#L927-L928), [Link2](https://github.com/ramyananth/servo/blob/master/components/script/dom/window.rs#L945-L946), [Link3](https://github.com/ramyananth/servo/blob/master/components/script/dom/workerglobalscope.rs#L415-L416), [Link4](https://github.com/ramyananth/servo/blob/master/components/script/dom/workerglobalscope.rs#L433-L434).
We tried to set imageBitmap.origin_clean = origin_is_clean on but it threw a private variable access error.

The current implementation throws the following error(s),

```
error[E0596]: cannot borrow data in a dereference of `dom::bindings::root::Root<dom::bindings::root::Dom<dom::imagebitmap::ImageBitmap>>` as mutable
   --> components/script/dom/window.rs:927:21
    |
927 |                     imageBitmap.set_bitmap_data(data);
    |                     ^^^^^^^^^^^ cannot borrow as mutable
    |
    = help: trait `DerefMut` is required to modify through a dereference, but it is not implemented for `dom::bindings::root::Root<dom::bindings::root::Dom<dom::imagebitmap::ImageBitmap>>`

```
Can you suggest a method to resolve this issue and get our build to pass?
Thank you in advance.

jdm commented:
The lack of mutability is the reason for the Cell and DomRefCell types. You can read more about them at https://doc.rust-lang.org/stable/std/cell/struct.Cell.html and https://doc.rust-lang.org/stable/std/cell/index.html. You will need to use the `set` and `borrow_mut` methods respectively to mutate the contents of those types.

ramyananth commented:
Hi Josh,
We added HTMLCanvas and OffscreenCanvas types for the Canvas implementation of Image bitmap. Our builds are successful for our changes on the terminal and it passes the `./mach test-tidy` test as well. How do we write the unit test cases for what we have implemented? 
Thanks!

The build looks like this,
```
ramya@intern-t440p:~/servo$ ./mach test-tidy
 0:04.13 INFO Diffing old and new manifests /home/ramya/servo/tests/wpt/webgl/meta/MANIFEST.json
 0:04.75 INFO Diffing old and new manifests /home/ramya/servo/tests/wpt/mozilla/meta/MANIFEST.json
 0:04.91 INFO Diffing old and new manifests /home/ramya/servo/tests/wpt/metadata/MANIFEST.json
Checking the config file...
Checking the wpt manifest file...
Checking directories for correct file extensions...
Checking files for tidiness...
  Progress: 100% (4/4)
tidy reported no errors.
rustfmt 1.4.13-nightly (c126730 2020-03-31)
clang-format not installed. Skipping CPP formatting.
ramya@intern-t440p:~/servo$ ./mach build --dev
   Compiling script v0.0.1 (/home/ramya/servo/components/script)
   Compiling mozjs_sys v0.68.1 (https://github.com/servo/mozjs?rev=d4370798ee17ad16d52ff2e83c9055d19a98b26f#d4370798)
   Completed script v0.0.1 custom-build (run) in 8.4s
   Completed mozjs_sys v0.68.1 custom-build (run) in 19.1s
   Compiling mozjs v0.13.0 (https://github.com/servo/rust-mozjs#ea10bed2)
   Completed mozjs v0.13.0 custom-build (run) in 2.0s
   Completed mozjs v0.13.0 in 3.2s
   Completed mozjs_sys v0.68.1 in 6.3s
   Compiling layout_thread_2013 v0.0.1 (/home/ramya/servo/components/layout_thread)
   Compiling libservo v0.0.1 (/home/ramya/servo/components/servo)
   Completed libservo v0.0.1 in 18.8s
   Completed layout_thread_2013 v0.0.1 in 73.0s
   Completed script v0.0.1 in 275.8s
   Compiling servo v0.0.1 (/home/ramya/servo/ports/glutin)
   Completed servo v0.0.1 bin "servo" in 49.1s
    Finished dev [unoptimized + debuginfo] target(s) in 5m 52s
Build Completed in 0:05:52
```

ramyananth commented:
Also, quick question. Could you provide us with a few materials to test our implementation?
We looked at these links - [Link1](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/createImageBitmap) and [Link2](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/canvas) but are unsure about where and how we can use this to test our work. Thanks a lot for your help.

jdm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 37e026db073e3466304b1e9367605bf66343b18c with merge c2d608346e0ec50521bd9d7569c1c1340f9249fe...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/f9H-UYDHT3WO2EEIDJcIcg)

jdm commented:
If you run `./mach test-wpt tests/wpt/web-platform-tests/offscreen-canvas` you will observe a bunch of crashes because if the use of `unwrap()` on the width and height values that are passed in as optional members inside the dictionary.

Another change I suggest is merging the two duplicate implementations in a method in GlobalScope, which you can call from Window and WorkerGlbalScope by calling `self.upcast::<GlobalScope>().create_image_bitmap()`.

jdm commented on components/script/dom/globalscope.rs:
```diff
@@ -2227,6 +2231,66 @@ impl GlobalScope {
+        let global = self.global();
+        let in_realm_proof = AlreadyInRealm::assert(&global);
+        let p = Promise::new_in_current_realm(&global, InRealm::Already(&in_realm_proof));
+        if options.resizeWidth.unwrap() == 0 {
```
The spec says "is present and is 0", which means we should use `options.resizeWidth.map_or(false, |w| w == 0)`, rather than panicking if the resizeWidth value is not defined in JS. See the docs for [map_or](https://doc.rust-lang.org/stable/std/option/enum.Option.html#method.map_or). This should be used for the resizeHeight check too.

ramyananth replied:
This is a better way to do it. Thanks. The panicking does not take place anymore.

jdm commented on components/script/dom/globalscope.rs:
```diff
@@ -2227,6 +2231,66 @@ impl GlobalScope {
+        image: ImageBitmapSource,
+        options: &ImageBitmapOptions,
+    ) -> Rc<Promise> {
+        let global = self.global();
```
We can use `self` anywhere we use `global` now instead, and remove this line.

ramyananth replied:
Done.

jdm commented on components/script/dom/globalscope.rs:
```diff
@@ -2227,6 +2231,66 @@ impl GlobalScope {
+                }
+                p
+            },
+            _ => p,
```
Let's use `p.reject_error(Error::NotSupported)` so we don't cause timeouts in tests that will otherwise endlessly wait for something that will never happen.

ramyananth replied:
Added this and then returned p.

jdm commented on components/script/dom/imagebitmap.rs:
```diff
@@ -39,6 +42,14 @@ impl ImageBitmap {
     }
+
+    pub fn set_bitmap_data(&self, data: DomRefCell<Vec<u8>>) {
+        *self.bitmap_data.borrow_mut() = data.borrow_mut().to_vec();
```
This should take a `Vec<u8>` argument instead, and use `*self.bitmap_data.borrow_mut() = data;`

ramyananth replied:
Made this change and on the calling function as well.

jdm commented on components/script/dom/globalscope.rs:
```diff
@@ -2227,6 +2231,66 @@ impl GlobalScope {
+            ImageBitmapSource::HTMLCanvasElement(ref canvas) => {
+                // https://html.spec.whatwg.org/multipage/#check-the-usability-of-the-image-argument
+                if !canvas.is_valid() {
+                    p.reject_error(Error::InvalidState)
```
We're going to want to return p at this point, otherwise we'll still fetch the canvas data and update the image bitmap contents.

ramyananth replied:
That's clear. Got it.

jdm commented on components/script/dom/globalscope.rs:
```diff
@@ -2227,6 +2231,66 @@ impl GlobalScope {
+            ImageBitmapSource::OffscreenCanvas(ref canvas) => {
+                // https://html.spec.whatwg.org/multipage/#check-the-usability-of-the-image-argument
+                if !canvas.is_valid() {
+                    p.reject_error(Error::InvalidState)
```
We need to return p here as well.

ramyananth replied:
Done

jdm commented on components/script/dom/globalscope.rs:
```diff
@@ -2227,6 +2231,66 @@ impl GlobalScope {
+        }
+
+        if options.resizeHeight.unwrap() == 0 {
+            p.reject_error(Error::InvalidState)
```
These early errors should return p immediately to avoid running more code in an error state.

ramyananth replied:
got it

jdm commented on components/script/dom/globalscope.rs:
```diff
@@ -2227,6 +2231,66 @@ impl GlobalScope {
+            p.reject_error(Error::InvalidState)
+        }
+
+        let image_bitmap = ImageBitmap::new(&global, 0, 0).unwrap();
```
We're creating a 0x0 image bitmap here. We should wait to create the ImageBitmap until we know the size we need to create.

ramyananth replied:
We've set it to the canvas image's width and height. The image bitmap now gets created after the options.resize tests for width and height.

jdm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 5d641a507108104b0372c1da4da860524007733d with merge 6d93ea1ff0fba702eeaba944cd4d58a508141a56...

ramyananth commented:
Hi Josh, 
We ran `./mach test-wpt tests/wpt/web-platform-tests/offscreen-canvas --log-raw wpt.log`
and saw the following result,
```
  ▶ Unexpected subtest result in /offscreen-canvas/convert-to-blob/offscreencanvas.convert.to.blob.w.html:
  └ PASS [expected FAIL] Test that call convertToBlob on a OffscreenCanvas with tainted origin throws exception in a worker

  ▶ OK [expected ERROR] /offscreen-canvas/filter/offscreencanvas.filter.w.html

  Main thread got signal
  [992/1566] /offscreen-canvas/path-objects/2d.path.quadraticCurveTo.basic.worker.html
             /offscreen-canvas/path-objects/2d.path.quadraticCurveTo.scaled.worker.html
             /offscreen-canvas/path-objects/2d.path.moveTo.newsubpath.worker.html
             /offscreen-canvas/path-objects/2d.path.moveTo.multiple.worker.html
mach interrupted by signal or user action. Stopping.
```
Should we run `./mach update-wpt wpt.log` and push our changes? 
Ref: [Documentation](https://github.com/servo/servo/blob/master/tests/wpt/README.md#updating-the-upstream-tests)

Thank you.

JayalakshmiV commented:
Hi Josh, 
We ran `./mach test-wpt tests/wpt/web-platform-tests/offscreen-canvas --log-raw wpt.log`
and saw the following result,

```
Unexpected subtest result in /offscreen-canvas/convert-to-blob/offscreencanvas.convert.to.blob.w.html:
  PASS [expected FAIL] Test that call convertToBlob on a OffscreenCanvas with tainted origin throws exception in a worker
  OK [expected ERROR] /offscreen-canvas/filter/offscreencanvas.filter.w.html
  Main thread got signal
  [992/1566] /offscreen-canvas/path-objects/2d.path.quadraticCurveTo.basic.worker.html
             /offscreen-canvas/path-objects/2d.path.quadraticCurveTo.scaled.worker.html
             /offscreen-canvas/path-objects/2d.path.moveTo.newsubpath.worker.html
             /offscreen-canvas/path-objects/2d.path.moveTo.multiple.worker.html
mach interrupted by signal or user action. Stopping.
```

Should we run `./mach update-wpt wpt.log` and push our changes? 
Ref: [Documentation](https://github.com/servo/servo/blob/master/tests/wpt/README.md#updating-the-upstream-tests)

Thank you.

jdm commented:
There are a _lot_ of tests that have new results. In particular, there are a bunch of tests that previously were skipped because the createImageBitmap API was missing that now fail because it's present and we reject non-canvas image sources. That's expected behaviour, so you should do the following:
1. `wget https://community.taskcluster-artifacts.net/KqQWr_VwQDe55wGO_T193Q/0/public/test-wpt.log https://community.taskcluster-artifacts.net/PhPboOqiSa-gdhzBMJbL_w/0/public/test-wpt.log https://community.taskcluster-artifacts.net/Fzvx31ovS6GPF4eGGBl8rA/0/public/test-wpt.log https://community.taskcluster-artifacts.net/EfQsJZaXTBSTmct0hy3scw/0/public/test-wpt.log`
1. `./mach update-wpt test-wpt.log test-wpt.log.1 test-wpt.log.2 test-wpt.log.3` (or whatever the files end up being called)
1. `git add tests/wpt/metadata tests/wpt/webgl/meta`

Then commit the results.

jdm commented:
You can go ahead and [squash](https://blog.carbonfive.com/2017/08/28/always-squash-and-rebase-your-git-commits/) all the commits into one! Thanks!

ramyananth commented:
We are seeing three merge conflicts in the test files. Is there something that we can do to resolve it? Thanks.

jdm commented:
If you undo the changes to those files, that should avoid the issue. they are intermittent failures that occur on our CI, not related to the changes in this PR.

ramyananth commented:
In that case, should we undo changes to those files and git rebase once again?

jdm commented:
that would be great.

ramyananth commented:
resolved conflicts and merged commits to one!

bors-servo commented:
:boom: Test timed out

jdm commented:
@bors-servo r+
Thanks!

bors-servo commented:
:pushpin: Commit f5c33fd has been approved by `jdm`

<!-- @bors-servo r=jdm f5c33fd5af9c1e271eee227a9cc5112e92c465d2 -->

bors-servo commented:
:hourglass: Testing commit f5c33fd5af9c1e271eee227a9cc5112e92c465d2 with merge 9e98f86917d0e282b608e0ac960ecb963b7d32c5...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/ZPgMBBjgTxijI7PtkgE2Gg)

CYBAI commented:
@bors-servo retry

- #25513

bors-servo commented:
:hourglass: Testing commit f5c33fd5af9c1e271eee227a9cc5112e92c465d2 with merge 3e8e6d2adb03abee18c3755a0ecf98d09de2f864...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/Ge6eHgjeQSy1oe7GI93VMw)

jdm commented:
@bors-servo retry
* #25513

bors-servo commented:
:lock: Merge conflict

bors-servo commented:
:umbrella: The latest upstream changes (presumably #26351) made this pull request unmergeable. Please resolve the merge conflicts.

ramyananth commented:
Hi Josh, we have resolved the merge conflicts and have pushed the changes. Please let us know the next steps with respect to this pull request.
Thank you so much for your help!

jdm commented:
Looks great; thanks!
@bors-servo r+

bors-servo commented:
:pushpin: Commit ef6f99d has been approved by `jdm`

<!-- @bors-servo r=jdm ef6f99d8f52549e40cee25ead7fabdf04c5bcc54 -->

jdm commented:
@bors-servo r-
Spoke too soon.

jdm commented:
Nope, it's all good.
@bors-servo r+

bors-servo commented:
:pushpin: Commit ef6f99d has been approved by `jdm`

<!-- @bors-servo r=jdm ef6f99d8f52549e40cee25ead7fabdf04c5bcc54 -->

bors-servo commented:
:hourglass: Testing commit ef6f99d8f52549e40cee25ead7fabdf04c5bcc54 with merge 2006ddabcdc5646b23e6dd298ce48a92b6793138...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/WRj18VmwQzSDx94oggSE9g)

jdm commented:
@bors-servo retry
* #26371

bors-servo commented:
:hourglass: Testing commit ef6f99d8f52549e40cee25ead7fabdf04c5bcc54 with merge 39e4eb43c383ef7312211937d98ace9935d50a43...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://community-tc.services.mozilla.com/tasks/groups/U82q_mMRSpCHZc4TrWLECg)
Approved by: jdm
Pushing 39e4eb43c383ef7312211937d98ace9935d50a43 to master...
